### ISSUE TITLE
How to know the Rank model improves the output of the Retrieval model in the case of MovieLens dataset?

### ISSUE THREAD

**Issue by houghtonweihu** [tensorflow/recommenders#690]
We all know that the rank will enhance the output of the retrieval model. But how can we see that in the case of MovieLens dataset?

**Comment by rlcauvin** [comment]
What do you mean by "enhance the output"? Do you want to compare the metrics (e.g. top K categorical accuracies) for the combined output of retrieval and ranking?

**Comment by rlcauvin** [comment]
We generally expect ranking models to be more predictive. However, it might depend on the metric. Using ROC-AUC as the evaluation metric, you will very likely see better predictions from the ranking model than from the retrieval model alone.

**Comment by rlcauvin** [comment]
The tutorials split the input data into training and test samples. The test sample includes the "actual" or "ground truth" values.

After building the retrieval and ranking models, generate predictions on the test data as you've described by invoking the retrieval model first, then the ranking model on initial recommendations from the retrieval model. The output gives you the "predicted" values.

You may then compute the [top K categorical accuracy](tf.keras.metrics.top_k_categorical_accuracy) and/or [ROC-AUC](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/AUC) on the actual and predicted values.

**Comment by rlcauvin** [comment]
If you do what I described above (compute evaluation metrics for combined retrieval and ranking), as well as the metrics for retrieval alone, you can compare them on the test sample. As for my own content recommendations models, I've found that ranking improves the ROC-AUC but not the top K categorical accuracies.

**Comment by rlcauvin** [comment]
I'm gathering that you want sort of a real-world before and after comparison? The positive engagement rate was X when using retrieval for recommendations, and it changed to Y after adding ranking in the recommendations? If that's what you're asking, no, I'm not aware of such an example.

**Comment by rlcauvin** [comment]
If you want to exclude watched movies from the recommendations, use [query_with_exclusions](https://www.tensorflow.org/recommenders/api_docs/python/tfrs/layers/factorized_top_k/BruteForce#query_with_exclusions) for retrieval, then do the ranking.

**Comment by patrickorlando** [comment]
Hi @houghtonweihu,
@rlcauvin is correct. In the case of the Movielens dataset, you would expect the ranking model to result in better performance on your chosen metrics. Like ROC-AUC, Hitrate@K (top K categorical accuracy), Precision@K, Recall@K.
